compaction/compaction_manager.cc

```diff
@@ -1185,6 +1185,33 @@ future<> compaction_manager::perform_sstable_scrub_validate_mode(replica::table*
     return perform_task(seastar::make_shared<validate_sstables_compaction_task>(*this, t, std::move(all_sstables)));
 }
 
+future<> compaction_manager::maybe_wait_on_current_compaction(replica::table* t) {
+    auto sstables = boost::copy_range<std::vector<sstables::shared_sstable>>(
+            *(t->get_sstables()) |
+            boost::adaptors::filtered([](const sstables::shared_sstable& s) { return s->get_origin() == "memtable"; }));
+    sstables = get_non_compacting_sstables(sstables);
+    if (sstables.size() < t->schema()->max_compaction_threshold() || t->is_auto_compaction_disabled_by_user()) {
+        co_return;
+    }
+    auto tasks_compacting_table = boost::copy_range<std::vector<shared_ptr<task>>>(_tasks |
+            boost::adaptors::filtered([table=t](const shared_ptr<task>& t) { return t->compacting_table() == table; }));
+    if (tasks_compacting_table.empty()) {
+        co_return;
+    }
+
+    // Wait until any of the currently running compactions on the table finish.
+    auto s = seastar::make_lw_shared<std::optional<seastar::promise<>>>(seastar::promise<>());
```

Comment:
There's no reason for the pointer if we're in a coroutine, a local would work just as well.

Reply:
The function will exit on first completed task and the continuation would then still try to access the optional promise leading to use after free (if we have multiple compactions on the same table).